tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/field_access.bal

```diff
@@ -353,10 +353,72 @@ function testFieldAccessOnMapConstruct() returns boolean {
     return "Sanjiva" == name;
 }
 
+type FooRec1 record {
+    json j;
+};
+
+type FooRec2 record {
+    json j;
+    FooRec1 fooRec;
+};
+
+function testFieldAccessOnJsonTypedRecordFields() returns error? {
+    FooRec1 rec1 = {j: "1"};
+    string val = check rec1.j;
+    assertEquals(val, "1");
+
+    rec1 = {j: {k: "2"}};
+    val = check rec1.j.k;
+    assertEquals(val, "2");
+
+    rec1 = {j: {k: {l: "1", m: "3"}}};
+    val = check rec1.j.k.m;
+    assertEquals(val, "3");
+
+    FooRec2 rec2 = {j: "1", fooRec: {j: "2"}};
+    val = check rec2.j;
+    assertEquals(val, "1");
+    val = check rec2.fooRec.j;
+    assertEquals(val, "2");
+
+    rec2 = {j: "1", fooRec: {j: {k: "3", l: {m: "4"}}}};
+    val = check rec2.fooRec.j.k;
+    assertEquals(val, "3");
+    val = check rec2.fooRec.j.l.m;
+    assertEquals(val, "4");
+}
+
+function testFieldAccessOnJsonTypedRecordFieldsResultsError() {
+    FooRec1 rec1 = {j: "1"};
+    json|error val = rec1.j.k.l;
+    assertEquals((<error>val).message(), "{ballerina}JSONOperationError");
```

Comment:
Please assert the "message" field from the error detail too. That has more information.